libr/anal/sign.c

```diff
@@ -67,6 +67,44 @@ R_API RList *r_sign_fcn_vars(RAnal *a, RAnalFunction *fcn) {
 	return ret;
 }
 
+R_API RList *r_sign_fcn_types(RAnal *a, RAnalFunction *fcn) {
+
+	// From anal/types/*:
+	// Get key-value types from sdb matching "func.%s", fcn->name
+	// Get func.%s.args (number of args)
+	// Get type,name pairs
+	// Put everything in RList following the next format:
+	// types: main.ret=%type%, main.args=%num%, main.arg.0="int,argc", ...
+
+	r_return_val_if_fail (a && fcn, NULL);
+
+	RList *ret = r_list_newf ((RListFree) free);
+	SdbKv *kv = NULL;
+	SdbListIter *it = NULL;
+	char *args_expr = r_str_newf ("func.%s.args", fcn->name);
+	SdbList *fcntypes = sdb_foreach_match (a->sdb_types, args_expr, 0);
+
+	if (fcntypes) {
+		ls_foreach (fcntypes, it, kv) {
+			if (strcmp (sdbkv_key (kv), args_expr) == 0) {
+				char *ret_type = sdb_get (a->sdb_types, r_str_newf ("func.%s.ret", fcn->name), 0);
```

Comment:
The `r_str_newf ("func.%s.ret"` is leaked. Since that string is used multiple times, it makes sense to allocate it once outside the loop, just like `args_expr` and free it at the end of the function.

The returned value of `sdb_get()` is leaked too. Here you can just use `sdb_const_get()` instead of regular get to fix it.